### PR TITLE
Tracks: Opt Out

### LINKED ISSUE #222: Tracks: Opt Out
### Details:
- [x] Add a new boolean field to the`Settings` entity (Simperium backed) to persist if the user opted out from Tracks
- [x] Add a new Switch in the Settings UI

### PR DESCRIPTION
**Pull request by jleandroperez**
### Notes:
**ALL** of the apps must assume the Preferences Object's Key will be `preferences-key`. That way we can ensure there'll be just **one** entity of that kind (same mechanism used for the Welcome Note!).

### Details:
- Options UI: New Tracks Opt-Out Switch
- New Data Model Mark 4
- Podfile.lock updated (as per Cocoapods 1.5.3)
- New Preferences Entity
- New SPCredentials property (SPSimperiumPreferencesObjectKey)

Closes #222
@roundhill ready for review sir!! thanks in advance!!

### Scenario: Sync'ing between two instances
1. Check out commit [4895b76](
https://github.com/Automattic/simplenote-ios/pull/228/commits/4895b765fc19046d69e6bbf51362b08b47f16d7f)
2. Run Simplenote in the **Simulator** + In **Your device**. (From now on we'll call them Instance A / Instance B)
3. Log into your account, in both instances
4. **Instance A:** Open the App's Setting and **verify**  the Analytics switch is On by default
5. **Instance A:** Flip the switch to disable Analytics
6. **Instance B:**  Open the same Settings screen. Verify that the Analytics are off

### Scenario: Analytics Off
1. Check out commit [4895b76](
https://github.com/Automattic/simplenote-ios/pull/228/commits/4895b765fc19046d69e6bbf51362b08b47f16d7f)
2. Log into your account
3. Open the App's Setting and **disable** analytics
4. Open the Note's list and press over any  note
5. Verify the following shows up in the console:  `## Tracking TRACKS OFF: list_note_opened`
6. Re-enable the Analytics Switch
7. Verify the following shows up in the console: `## Tracking TRACKS ON: list_note_opened`

### CODE REVIEW

**Comment by roundhill** [comment]
`Preferences.h` and `Preferences.m` appear to be not linked to the project:

screen shot 2018-09-18 at 12 48 52 pm

**Comment by roundhill** [comment]
> SPSimperiumPreferencesObjectKey = preferences-key

Why is this in config.plist? Can't it just be a constant or something?

**Comment by jleandroperez** [comment]
@roundhill Dan! all changes UP:

- Nuked the new `config.plist` setting. It's now a constant: `preferences-key`
- Privacy Settings now has it's own ViewController
- Footer Implemented. On tap it'll open Safari

Ready for another round!! thanks in advance!!!

File: Simplenote.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -290,6 +293,9 @@
...
 		B59314DC1A486C0100B651ED /* VSTheme+Simplenote.h */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.c.h; name = "VSTheme+Simplenote.h"; path = "Classes/VSTheme+Simplenote.h"; sourceTree = "<group>"; };
 		B59314DD1A486C0100B651ED /* VSTheme+Simplenote.m */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.c.objc; name = "VSTheme+Simplenote.m"; path = "Classes/VSTheme+Simplenote.m"; sourceTree = "<group>"; };
+		B594E60C21508170001577EE /* Simplenote 4.xcdatamodel */ = {isa = PBXFileReference; lastKnownFileType = wrapper.xcdatamodel; path = "Simplenote 4.xcdatamodel"; sourceTree = "<group>"; };
+		B594E61321508247001577EE /* Preferences.m */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.c.objc; name = Preferences.m; path = "/Users/lantean/Developer/Automattic/simplenote-ios/Simplenote/Preferences.m"; sourceTree = "<absolute>"; };
+		B594E61421508247001577EE /* Preferences.h */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.c.h; name = Preferences.h; path = "/Users/lantean/Developer/Automattic/simplenote-ios/Simplenote/Preferences.h"; sourceTree = "<absolute>"; };
```
Comment (roundhill) [inline Simplenote.xcodeproj/project.pbxproj]: Uh oh!

**Comment by roundhill** [comment]
Looking good, but we still need to link the `Learn more` text to https://automattic.com/cookies.

**Comment by roundhill** [comment]
Also can you check on why Simperium is getting a `1` and `0` instead of the JSON `true` or `false` string value from Simperium iOS? It'd make it easier on the other apps if we didn't have to do an int check.

**Comment by jleandroperez** [comment]
Thanks for reviewing this Dan!!. The `Learn More` is "clickable", i'll add an underline to make it extremely explicit (or blue!).

Regarding the 0/1, i've debugged the `deleted` flag in the **Note** entity. This is the diff that Simplenote (iOS) sends, whenever a Note is deleted:

```
2018-09-20 00:38:26.054940-0300 Simplenote[51059:4171734] Simperium entity diff found 3 changed members
(lldb) po changes
<__NSArrayM 0x6000038b5440>(
{
    ccid = 31d1db7650604dbf86d34fc13c393152;
    id = 455d35c12e5f477d809d2d097ff62aa5;
    o = M;
    sv = 1;
    v =     {
        deleted =         {
            o = r;
            v = 1;
        };
        modificationDate =         {
            o = r;
            v = "1537414606.300598";
        };
        tags =         {
            o = r;
            v =             (
            );
        };
    };
}
)
```

Internally, Booleans are represented by means of Integers. I suspect this has been working this way for a long, really really long time. (I'm trying to remember), i think there was a way to force the field type by applying a Schema, but i can't spot in my notes how that worked!

**Comment by alaskaa30** [comment]
👍

Le 19 sept. 2018 11:41 p.m., "Jorge Leandro Perez" <notifications@github.com>
a écrit :

Thanks for reviewing this Dan!!. The Learn More is "clickable", i'll add an
underline to make it extremely explicit (or blue!).

Regarding the 0/1, i've debugged the deleted flag in the *Note* entity.
This is the diff that Simplenote (iOS) sends, whenever a Note is deleted:

2018-09-20 00:38:26.054940-0300 Simplenote[51059:4171734] Simperium
entity diff found 3 changed members
(lldb) po changes
<__NSArrayM 0x6000038b5440>(
{
    ccid = 31d1db7650604dbf86d34fc13c393152;
    id = 455d35c12e5f477d809d2d097ff62aa5;
    o = M;
    sv = 1;
    v =     {
        deleted =         {
            o = r;
            v = 1;
        };
        modificationDate =         {
            o = r;
            v = "1537414606.300598";
        };
        tags =         {
            o = r;
            v =             (
            );
        };
    };
}
)

Internally, Booleans are represented by means of Integers. I suspect this
has been working this way for a long, really really long time. (I'm trying
to remember), i think there was a way to force the field type by applying a
Schema, but i can't spot in my notes how that worked!

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/Automattic/simplenote-ios/pull/228#issuecomment-423030180>,
or mute the thread

**Comment by jleandroperez** [comment]
### Update:
I've ended up replicated WPiOS's UI (screenshots below, in Dark / Light mode).
Reason: the tableView footer wasn't accepting formatted text, and this seemed like a (way) nicer alternative!.

Ready for another look, thanks!!

simulator screen shot - iphone se - 2018-09-20 at 01 16 10
simulator screen shot - iphone se - 2018-09-20 at 01 16 14

**Comment by roundhill** [comment]
> and this seemed like a (way) nicer alternative!.

Beautiful! ❤️

**Comment by roundhill** [comment]
> Regarding the 0/1, i've debugged the `deleted` flag in the **Note** entity. This is the diff that Simplenote (iOS) sends, whenever a Note is deleted:

So here's where the Android app handles getting the deleted property:

https://github.com/Automattic/simplenote-android/blob/develop/Simplenote/src/main/java/com/automattic/simplenote/models/Note.java#L329-L338

So it looks like it is adjusting for either a boolean value or an integer... seems like we should fix this in simperium-ios? I will check into the schema thing.

**Comment by roundhill** [comment]
> So it looks like it is adjusting for either a boolean value or an integer... seems like we should fix this in simperium-ios? I will check into the schema thing.

After further research, it looks like this is `the way it is` for now, we'll need to make sure the other clients check for the integer as well as the boolean.

**Comment by jleandroperez** [comment]
Thanks a lot Dan!!!!!!

**Pull request merged**